fix: user created after breaking when the user creation date is nil

andercampos opened the pull request:
# 🐞 The problem

We had an incident when using the userCreatedAfter strategy, as can be seen in this [slack thread](https://pipefy.slack.com/archives/GEY4N54LQ/p1684529970445339). It happened only in the public forms because the `user_created_at` parameter was null in this scenario, and the code was not prepared for this scenario.

# 💻 The fix

- Add validation for the user creation date

# 🗂 Related cards

[Fix the bug of the userCreatedAfter Unleash strategy](https://app.pipefy.com/open-cards/716329887) from [Activation - PLG pipe](https://app.pipefy.com/pipes/302207022)

# 📑 For the reviewer

- To be compliant with SOC and ISO policies, all MRs need to be associated with an open card, which must be linked to on their descriptions. Please, do not approve or merge MRs without a card link attached.

- Check out our [Guidelines for the Reviewer](https://git.pipefy.net/pipefy/guidelines/-/blob/master/docs/code-review/for-the-reviewer/README.md) to assist your review process.

Commit:
fix: user created after breaking when the user creation date is nil

gbrlcustodio commented on lib/unleash/strategy/user_created_after.rb:
```diff
@@ -15,9 +15,12 @@ def is_enabled?(params = {}, context = nil)
         return false unless context.instance_of?(Unleash::Context)
         return false if forbidden_org_uuids(params).include?(current_org_uuid(context))
 
+        user_creation_date = user_created_at(context)
++       return false if user_creation_date.nil?
```
What if, instead of parsing nil, we're parsing a non-parseable datetime, like `non-parseable`, `1` , etc? Should we cover this scenario as well?

GiovanaValadao replied:
Sorry for the delay 😅 good point! I converted the date to a string so it will always be a valid object to parse, but, if the value is not appropriate, it will fall in the `rescue ArgumentError`. I don't know if it is the best solution but I think it does the trick 😬  WDYT @gbrlcustodio ?

gbrlcustodio replied:
Welp, it doesn't seem to be enough for covering mentioned scenarios, whenever I try parsing inputs like `''` and `'1'`, I get a `Date::Error` (see in the examples below). Do we have any testing scenarios covering those?

```ruby
irb(main):005:0> DateTime.parse('1')
Traceback (most recent call last):
        ...
        1: from (irb):5:in `parse'
Date::Error (invalid date)
irb(main):006:0> DateTime.parse('')
Traceback (most recent call last):
        ...
        1: from (irb):6:in `parse'
Date::Error (invalid date)
```

Am I doing something wrong?

gbrlcustodio replied:
Wouldn't rescuing for such errors and returning false as a safeguard be easier in this scenario? What's your opinion?

GiovanaValadao replied:
So, the `Date::Error` inherits from `ArgumentError` (which is the one being rescued) so these scenarios are being covered 😸

gbrlcustodio replied:
That explains! Thanks for letting me know :1st_place_medal:

gbrlcustodio commented on spec/unleash/strategy/email_from_domain_spec.rb:
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require 'unleash/context'
+require 'unleash/strategy/email_from_domain'
```
Out of curiosity, were these missing `require` preventing specs from running?

GiovanaValadao replied:
Yes 😞

gbrlcustodio reviewed (commented):
Great job!

GiovanaValadao committed:
fix: Use to_s in the user_created_at

Use to_s in the user_created_at so that, any non-parseable value will fall under the ArgumentError

gbrlcustodio reviewed (approved):
LGTM! :1st_place_medal:
